lib/de.ml

```diff
@@ -2605,6 +2892,1001 @@ module Def = struct
+      Array.fill stats.new_observations 0 num_observation_types 0
+      ; Array.fill stats.observations 0 num_observation_types 0
+      ; stats.num_new_observations <- 0
+      ; stats.num_observations <- 0
+
+    let reset_symbol_frequencies c =
+      Array.fill c.freqs.litlen 0 (Array.length c.freqs.litlen) 0
+      ; Array.fill c.freqs.offset 0 (Array.length c.freqs.offset) 0
+
+    let finish_sequence seq litrunlen = seq.litrunlen_and_length <- litrunlen
+    let num_observations_per_block_check = 512
+
+    let do_end_block_check stats block_length =
+      let f () =
+        if stats.num_observations > 0 then (
+          let total_delta = ref 0 in
+          for i = 0 to num_observation_types - 1 do
+            let expected = stats.observations.(i) * stats.num_new_observations in
+            let actual = stats.new_observations.(i) * stats.num_observations in
+            let delta =
+              if actual > expected then actual - expected else expected - actual
+            in
+            total_delta := !total_delta + delta
+          done
+          ; if
+              !total_delta + (block_length / 4096 * stats.num_observations)
+              >= num_observations_per_block_check
+                 * 200
+                 / 512
+                 * stats.num_observations
+            then true
+            else false)
+        else false in
+      if f () then true
+      else (
+        for i = 0 to num_observation_types - 1 do
+          stats.num_observations <-
+            stats.num_observations + stats.new_observations.(i)
+          ; stats.observations.(i) <-
+              stats.observations.(i) + stats.new_observations.(i)
+          ; stats.new_observations.(i) <- 0
+        done
+        ; stats.num_new_observations <- 0
+        ; false)
+
+    let should_end_block stats in_block_begin in_next in_end =
+      if
+        stats.num_new_observations < num_observations_per_block_check
+        || in_next - in_block_begin < _min_block_length
+        || in_end - in_next < _min_block_length
+      then false
+      else do_end_block_check stats (in_next - in_block_begin)
+
+    let rec _lz_extend i start_pos match_pos len max_len =
+      if
+        len < max_len
+        && unsafe_get_uint8 i (match_pos + len)
+           = unsafe_get_uint8 i (start_pos + len)
+      then _lz_extend i start_pos match_pos (len + 1) max_len
+      else len
+
+    let rec lz_extend i start_pos match_pos len max_len =
+      if
+        max_len - len >= 4
+        && unsafe_get_uint32 i (match_pos + len)
+           = unsafe_get_uint32 i (start_pos + len)
+      then lz_extend i start_pos match_pos (len + 4) max_len
+      else _lz_extend i start_pos match_pos len max_len
+
+    let hc_matchfinder_slide_window mf =
+      Array.iteri
+        (fun i v -> mf.hash4_tab.(i) <- v - (window_size land 0xFFFF))
+        mf.hash4_tab
+      ; Array.iteri
+          (fun i v -> mf.next_tab.(i) <- v - (window_size land 0xFFFF))
+          mf.next_tab
```

Comment:
may be instead of the use of `Array.iteri`, the best is to use a simple `for` loop.